### PR TITLE
Add check_ops.assert_shapes and tests

### PR DESCRIPTION
**Pull request by bodin-e**
Issue #24779

Now supports static and dynamic shape checks. 

Names are now passed via tensors only. 

A suggestion included is also the 'event_shape_only' flag for when only checks on the rightmost dimensions are warranted.

### CODE REVIEW

File: tensorflow/python/kernel_tests/check_ops_test.py
Comment (alextp) [inline tensorflow/python/kernel_tests/check_ops_test.py]: TF style prefers to use parenthesis over \ to break lines

File: tensorflow/python/kernel_tests/check_ops_test.py
Comment (alextp) [inline tensorflow/python/kernel_tests/check_ops_test.py]: I think this non-eager second half is better written as its own separate test which explicitly enters graph mode with ops.Graph().as_default(). Same below.

File: tensorflow/python/ops/check_ops.py
Comment (alextp) [inline tensorflow/python/ops/check_ops.py]: Can you run the api compatibility test in tensorflow/tools/api/tests:api_compatibility_test and follow the instructions to regenerate the golden files?

File: tensorflow/python/ops/check_ops.py
Comment (alextp) [inline tensorflow/python/ops/check_ops.py]: Why is it named event_shape? Also the docstring is somewhat confusing for this method. When would a user want to set it to true vs false?

File: tensorflow/python/ops/check_ops.py
Comment (alextp) [inline tensorflow/python/ops/check_ops.py]: Do you need this? I think tf.control_dependencies is perfectly happy to take an empty list.

File: tensorflow/python/ops/check_ops.py
Comment (alextp) [inline tensorflow/python/ops/check_ops.py]: This can live in the toplevel

Comment (alextp) [reply tensorflow/python/ops/check_ops.py]: Same for the other inner functions below

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/check_ops.py
Comment (bodin-e) [reply tensorflow/python/ops/check_ops.py]: Setting this to true would be useful in scenarios where the user is writing a function that only uses (and have expectations on) the inner-most dimensions of the passed tensors. As when the function is to be broadcasted over any number of the remaining (preceding) dimensions.

An example of such a function is matmul; as where the inner-most two dimensions are to represent the matrix shape of respective operand, while the rest of the dimensions are broadcasted.

I take your point that it was confusing. The name event_shape came from the terminology in 'distributions': https://www.tensorflow.org/api_docs/python/tf/distributions/Normal#event_shape 
But, as this has a wider context, a better name to describe it could be 'innermost_dims', to be in line with terminology as in: https://www.tensorflow.org/api_docs/python/tf/linalg/matmul

However, after having another thought about this flag, the signature I show below could be cleaner and more intuitive. Rather than having a flag at all, this could be achieved meanwhile relying on standard numpy notation for familiar semantics:

Example:
tf.assert_shapes({
	x: (..., 'a', 'b'),
	y: (..., 'b', 'c')
})

Here the rule applied is that if the specified shape contains `...` as the first element, the remaining elements are checked as inner-most (relative to the end via negative indexing). I.e. the first dimensions up to the specified ones are ignored, similar to how `...` is used in numpy. Specifying `...` in any other entry is disallowed (raising a ValueError) to be in line with that semantic and allow for error messages pointing to specific dimensions (as e.g. 'dimension 0 must have size 3' or 'dimension -2 must have size 3).

What do you think?

Comment (alextp) [reply tensorflow/python/ops/check_ops.py]: I love this.

File: tensorflow/python/kernel_tests/check_ops_test.py
Comment (alextp) [inline tensorflow/python/kernel_tests/check_ops_test.py]: I was showing your operation to some other core TF devs and someone asked for a unit test that says that you can use any sequence, including a string, to validate shapes. So something like

tf.assert_shapes({
 image: "NCHW",
 kernel: "Cc",
})

would validate a conv kernel. How do you feel about this?

File: tensorflow/python/ops/check_ops.py
Comment (bodin-e) [reply tensorflow/python/ops/check_ops.py]: Great!

**Review by alextp — DISMISSED** [review]

File: tensorflow/python/kernel_tests/check_ops_test.py
Comment (bodin-e) [reply tensorflow/python/kernel_tests/check_ops_test.py]: I like that idea a lot. 

To enable the same capabilities using e.g. strings we would need a character symbol for `...` and None (*any* size). In the code just pushed it is currently `#` for `...` and  `*` for *any*.

In the code I made it an arbitrary iterable, then not needing the attribute `__len__` but only the `__iter__` and then use the tuple(iterable) constructor to guarantee having `__len__` for the specified shape rank checks.

As certain iterable/sequence types have their own element types, as e.g. `str` having `str` elements, we need to be cast/parse elements to integers for the explicit size check case. To handle it for general/unknown element types the pythonic way I know is a try-catch int parsing attempt. If you prefer another approach I'd be happy to do it differently.

What do you think?

Comment (alextp) [reply tensorflow/python/kernel_tests/check_ops_test.py]: I think the string case is good even if it's more limited (no None, number, or ellipsis).

Comment (bodin-e) [reply tensorflow/python/kernel_tests/check_ops_test.py]: I definitely think too the string case is good either way.

Although I think that it would still be neat to have all functionality in the string/general iterable case as well, as it would allow for a compact syntax as a complete alternative. 

If alternative symbols were to be included, perhaps they could instead by the following to have similar meaning as in regular expressions:
*any*: '.'
ellipsis: '*'

Examples:
```python
def matmul(x, y):
  tf.assert_shapes({
      x: (..., 'A', 'B'),
      y: (..., 'B', 'C')
  })
  ...
```
and as there is no need to declare 'A' and 'C' it could be written:	
```python
tf.assert_shapes({
  	x: (..., '.', 'B'),
  	y: (..., 'B', '.')
})
```
or using strings:
```python
tf.assert_shapes({
  x: "*.B",
  y: "*B."
})
```

Comment (bodin-e) [reply tensorflow/python/kernel_tests/check_ops_test.py]: Great. Pushed the symbol changes

**Comment by bodin-e** [comment]
Anything else I can do on this?

**Review by alextp — DISMISSED** [review]

**Comment by alextp** [comment]
@rthadur can we merge this?

**Comment by rthadur** [comment]
> @rthadur can we merge this?

sure

File: tensorflow/python/ops/check_ops.py
Comment (alextp) [inline tensorflow/python/ops/check_ops.py]: I think the ellipsis here is causing some of our python2 tests to fail with 

  File "tensorflow/python/ops/check_ops.py", line 1561
    return symbol in [..., '*']
SyntaxError: Type annotations and ellipsis literal not enabled.

Comment (bodin-e) [reply tensorflow/python/ops/check_ops.py]: I see. Apparently python2 does not support the Ellipsis literal, but it does have the Ellipsis type. So by replacing '...' by 'Ellipsis' in the symbol list it will be made backwards compatible with python2. I can push that change

Comment (alextp) [reply tensorflow/python/ops/check_ops.py]: Yes, thanks!

Comment (bodin-e) [reply tensorflow/python/ops/check_ops.py]: Pushed

**Comment by bodin-e** [comment]
I understand you must be busy with things relating to tf 2, but if there is anything I can do on this to speed things up let me know

**Comment by alextp** [comment]
@rthadur what is blocking this merge?

**Review by alextp — DISMISSED** [review]

**Comment by rthadur** [comment]
> @rthadur what is blocking this merge?

i started internal merge after approval , will update soon once it is done

**Comment by rthadur** [comment]
@bodin-e can you please check failed builds

**Comment by bodin-e** [comment]
> @bodin-e can you please check failed builds

Was a missing check for __getitem__ implementation of the shape dict entries, as an alternative to __iter__, as either is sufficient for the iter() call (made inside the tuple construction of tuple(shapes[x])). The python 2 string type does not implement __iter__ which caused the errors.

Fixed and pushed now.

**Review by alextp — DISMISSED** [review]

**Comment by rthadur** [comment]
@bodin-e can you please check build failures.

**Comment by bodin-e** [comment]
> @bodin-e can you please check build failures.

I think I had made a mistake with regards to the tf_export annotation. 

I have now annotated the v1 version with @tf_export(v1=['debugging.assert_shapes']). (I assume v1 is for graph mode as the other v1 versions of assert functions return the op explicitly)

And I have added a v2 version (assert_shapes_v2) with @tf_export('debugging.assert_shapes', v1=[]).

Is this the correct way?

**Comment by alextp** [comment]
@bodin-e yes this is how we handle different exports for v1 and v2.

**Review by alextp — DISMISSED** [review]

**Comment by bodin-e** [comment]
I will look into what differs in the golden as when the latest master is merged into my branch locally, to see the same diff as happening server-side.

**Comment by bodin-e** [comment]
> I will look into what differs in the golden as when the latest master is merged into my branch locally, to see the same diff as happening server-side.

There was a related difference in a golden that I had missed, added it now.

There were also some unrelated differences in the golden in master, with path "tensorflow.summary", see below. 

```javascript
E0426 20:57:09.711776 4620424640 api_compatibility_test.py:254] 1 differences found between API and golden.
Issue 1	:
  path: "tensorflow.summary"
  tf_module {
    member {
      name: "SummaryWriter"
      mtype: "<type \'type\'>"
    }
    member {
      name: "experimental"
      mtype: "<type \'module\'>"
+   }
+   member_method {
+     name: "audio"
+     argspec: "args=[\'name\', \'data\', \'sample_rate\', \'step\', \'max_outputs\', \'encoding\', \'description\'], varargs=None, keywords=None, defaults=[\'3\', \'None\', \'None\'], "
    }
    member_method {
      name: "create_file_writer"
      argspec: "args=[\'logdir\', \'max_queue\', \'flush_millis\', \'filename_suffix\', \'name\'], varargs=None, keywords=None, defaults=[\'None\', \'None\', \'None\', \'None\'], "
    }
    member_method {
      name: "create_noop_writer"
      argspec: "args=[], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "flush"
      argspec: "args=[\'writer\', \'name\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
+     name: "histogram"
+     argspec: "args=[\'name\', \'data\', \'step\', \'buckets\', \'description\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
+   }
+   member_method {
+     name: "image"
+     argspec: "args=[\'name\', \'data\', \'step\', \'max_outputs\', \'description\'], varargs=None, keywords=None, defaults=[\'3\', \'None\'], "
+   }
+   member_method {
      name: "record_if"
      argspec: "args=[\'condition\'], varargs=None, keywords=None, defaults=None"
+   }
+   member_method {
+     name: "scalar"
+     argspec: "args=[\'name\', \'data\', \'step\', \'description\'], varargs=None, keywords=None, defaults=[\'None\'], "
+   }
+   member_method {
+     name: "text"
+     argspec: "args=[\'name\', \'data\', \'step\', \'description\'], varargs=None, keywords=None, defaults=[\'None\'], "
    }
    member_method {
      name: "trace_export"
      argspec: "args=[\'name\', \'step\', \'profiler_outdir\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
      name: "trace_off"
      argspec: "args=[], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "trace_on"
      argspec: "args=[\'graph\', \'profiler\'], varargs=None, keywords=None, defaults=[\'True\', \'False\'], "
    }
    member_method {
      name: "write"
      argspec: "args=[\'tag\', \'tensor\', \'step\', \'metadata\', \'name\'], varargs=None, keywords=None, defaults=[\'None\', \'None\', \'None\'], "
    }
  }
```

**Review by alextp — APPROVED** [review]

**Pull request merged**
